### PR TITLE
David/file upload

### PR DESCRIPTION
**Pull request by ludavidca**
# Notion ticket link

[Implement File Uploading](https://www.notion.so/uwblueprintexecs/Implement-CRUD-for-Google-Drive-API-69b22dcc28f24256be626080e85e67f9?pvs=4)

## Implementation description
- Created simple UI for file uploading
- Implemented NEXTJS routes for uploading to google drives
- Created Google Project to facilitate Create operations

## Steps to test

1. Pull new env variables from vault
2. Navigate to localhost:3000/page and upload a file
3.  Check sistema@blueprint.org to ensure that the file is in the lesson plan folder

## What should reviewers focus on?
- Cleanliness of code
- Ability for team members to understand and build on top of existing code

## Checklist

- [x] My PR name is descriptive and in imperative tense
- [x] My commit messages are descriptive and in imperative tense. My commits are atomic and trivial commits are squashed or fixup'd into non-trivial commits
- [x] I have run the appropriate linter(s)
- [x] I have requested a review from the PL, as well as other devs who have background knowledge on this PR or who will be building on top of this PR

### CODE REVIEW

**Comment by ChinemeremChigbo** [comment]
Please look into the failing logs. Consider using [nektos](https://github.com/nektos/act) to run Github Actions locally (and not need to make a commit each time you want to test). This [StackOverflow link](https://stackoverflow.com/questions/56789424/the-incoming-json-object-does-not-contain-a-client-email-field) might be helpful for fixing the failing build issues. If you need any help with anything please let me know and we can do some pair programming:

**Review by ChinemeremChigbo — COMMENTED** [review]
Please look into the failing logs. Consider using [nektos](https://github.com/nektos/act) to run Github Actions locally (and not need to make a commit each time you want to test). This [StackOverflow link](https://stackoverflow.com/questions/56789424/the-incoming-json-object-does-not-contain-a-client-email-field) might be helpful for fixing the failing build issues. If you need any help with anything please let me know and we can do some pair programming:

**Review by phamkelly17 — APPROVED** [review]
lgtm, works great

**Review by ChinemeremChigbo — COMMENTED** [review]
Could you please address the merge conflicts

**Comment by ChinemeremChigbo** [comment]
> # Notion ticket link
> [Implement File Uploading] (https://www.notion.so/uwblueprintexecs/Implement-CRUD-for-Google-Drive-API-69b22dcc28f24256be626080e85e67f9?pvs=4)
> 
> ## Implementation description
> * Created simple UI for file uploading
> * Implemented NEXTJS routes for uploading to google drives
> * Created Google Project to facilitate Create operations
> 
> ## Steps to test
> 1. Pull new env variables from vault
> 2. Navigate to localhost:3000/page and upload a file
> 3. Check [sistema@blueprint.org](mailto:sistema@blueprint.org) to ensure that the file is in the lesson plan folder
> 
> ## What should reviewers focus on?
> * Cleanliness of code
> * Ability for team members to understand and build on top of existing code
> 
> ## Checklist
> * [ x ] My PR name is descriptive and in imperative tense
> * [ x ] My commit messages are descriptive and in imperative tense. My commits are atomic and trivial commits are squashed or fixup'd into non-trivial commits
> * [ x ] I have run the appropriate linter(s)
> * [ x ] I have requested a review from the PL, as well as other devs who have background knowledge on this PR or who will be building on top of this PR

To check a box in markdown, leave no spaces between the "x" and the brackets "[" "]" i.e.
* [x] like this

instead of
* [ x ] like this

File: src/pages/page.tsx [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+
+      console.log('Search result:', driveFiles);
+    } catch (error) {
+      console.error('Error searching files:', error);
+    }
+  };
+
+  useEffect(() => {
+    handleSearch({ preventDefault: () => {} });
+    // eslint-disable-next-line react-hooks/exhaustive-deps
```
Comment (ChinemeremChigbo) [inline src/pages/page.tsx:70]: Please don't disable eslint warnings unless you absolutely need to.

File: src/pages/page.tsx
[the remark this answers is not in the document]
Comment (ChinemeremChigbo) [reply src/pages/page.tsx:1]: Here is a modified version of this file that implements some of my comments. Please use it as reference.

```typescript
import { Container } from '@chakra-ui/react';
import { useCallback, useEffect, useState } from 'react';

type DriveFile = {
  id: string;
  name: string;
};

export default function FileUploadForm() {
  const [file, setFile] = useState<File | null>(null);
  const [fileName, setFileName] = useState('');
  const [driveFiles, setDriveFiles] = useState<DriveFile[]>([]);

  const handleFileChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    if (e.target.files) {
      setFile(e.target.files[0]);
      setFileName(e.target.files[0].name);
    }
  };

  const handleSubmit = async (e: React.FormEvent<HTMLFormElement>) => {
    e.preventDefault();
    if (!file) return;

    const formData = new FormData();
    formData.append('file', file);
    formData.append('fileName', fileName);

    try {
      const res = await fetch('/api/uploadFile/', {
        method: 'POST',
        body: formData,
      });

      const data = await res.json();
      if (res.ok) {
        await handleSearch();
        // Remove the deleted file from the state
        alert(data.message);
      } else {
        throw new Error(data.message);
      }
    } catch (error) {
      console.error('Error Uploading file:', error);
      alert('Error Uploading file');
    }
  };

  const handleSearch = useCallback(async () => {
    try {
      const res = await fetch('/api/searchDrive/');
      const data = await res.json();
      console.log('Search result:', data.results.files);
      setDriveFiles(data.results.files);
    } catch (error) {
      console.error('Error searching files:', error);
    }
  }, []);

  useEffect(() => {
    handleSearch();
  }, [handleSearch]);

  const handleFileDelete = async (fileId: string) => {
    try {
      const response = await fetch(`/api/deleteFile?fileId=${fileId}`, {
        method: 'DELETE',
      });
      const data = await response.json();

      if (response.ok) {
        await handleSearch();
        alert(data.message);
      } else {
        throw new Error(data.message);
      }
    } catch (error) {
      console.error('Error deleting file:', error);
      alert('Error deleting file');
    }
  };

  const handleFileClick = (fileId: string) => {
    window.open(`https://drive.google.com/file/d/${fileId}/view`, '_blank');
  };

  return (
    <Container>
      <form onSubmit={handleSubmit}>
        <input
          type="file"
          onChange={handleFileChange}
          accept="application/pdf"
        />
        <button type="submit">Upload</button>
      </form>
      <form onSubmit={handleSearch}>
        <button type="submit">Search</button>
      </form>
      <div className="Search Results">
        {driveFiles.map((file) => (
          <div key={file.id}>
            <button onClick={() => handleFileClick(file.id)}>
              {file.name + '_________'}
            </button>
            <button onClick={() => handleFileDelete(file.id)}>Delete</button>
            <br />
          </div>
        ))}
      </div>
    </Container>
  );
}

```

File: src/pages/page.tsx [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+
+    try {
+      const res = await fetch('/api/uploadFile/', {
+        method: 'POST',
+        body: formData,
+      });
+
+      const data = await res.json();
+      if (res.ok) {
+        const fakeEvent = { preventDefault: () => {} };
```
Comment (ChinemeremChigbo) [inline src/pages/page.tsx:35]: Using `fakeEvent` is unnecessary. You can directly call `handleSearch` without arguments if you refactor it not to rely on the event object.

**Review by ChinemeremChigbo — CHANGES_REQUESTED** [review]

File: .firebaserc [hunk 3]
```diff
@@ -0,0 +1,5 @@
+{
+  "projects": {
+    "default": "sistema-d720c"
```
[the remark this answers is not in the document]
Comment (ludavidca) [reply .firebaserc:3]: deleted

File: app/api/deleteFile/route.ts [hunk 4]
```diff
@@ -0,0 +1,48 @@
+import { google } from 'googleapis';
+import { NextRequest, NextResponse } from 'next/server';
+import { stringify } from 'querystring';
```
[the remark this answers is not in the document]
Comment (ludavidca) [reply app/api/deleteFile/route.ts:3]: done

File: app/api/uploadFile/route.ts [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+      private_key: private_key,
+    },
+  });
+  const drive = google.drive({ version: 'v3', auth });
+  try {
+    const googleRes = await drive.files.create({
+      requestBody: {
+        name: filename,
+        mimeType: 'application/pdf',
+        parents: ['1schkzvm_b46UGovHpQ2uH-X-nJtlm32_'],
```
[the remark this answers is not in the document]
Comment (ludavidca) [reply app/api/uploadFile/route.ts:31]: It does not matter as the id will not be accessible without the google drive env variables (these come from the links of the folder, and you cannot access said folder without being logged in)

File: src/pages/page.tsx [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+  const [file, setFile] = useState(null);
+  const [fileName, setFileName] = useState('');
+  const [driveFiles, setDriveFiles] = useState<DriveFile[]>([]);
+
+  const handleFileChange = (e) => {
+    setFile(e.target.files[0]);
+    setFileName(e.target.files[0].name);
+  };
+
+  const handleSubmit = async (e) => {
```
[the remark this answers is not in the document]
Comment (ludavidca) [reply src/pages/page.tsx:19]: will do in future

File: env.d.ts [hunk 7]
```diff
@@ -0,0 +1,16 @@
+declare global {
```
[the remark this answers is not in the document]
Comment (ludavidca) [reply env.d.ts:1]: You're right, I think it's mostly there for typesafety in env variables. I ran into some weird issues before adding this file. Is it ok if I leave it in for now?

File: app/api/uploadFile/route.ts [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+        driveId: '1schkzvm_b46UGovHpQ2uH-X-nJtlm32_',
+      },
+      media: {
+        mimeType: 'application/pdf',
+        body: Readable.from(fileBuffer),
+      },
+      supportsAllDrives: true,
+    });
+    console.log(googleRes);
```
[the remark this answers is not in the document]
Comment (ludavidca) [reply app/api/uploadFile/route.ts:40]: Nope. Removed

File: .env.sample [hunk 9]
```diff
@@ -6,11 +6,12 @@ EMAIL_PASS=
 EMAIL_SERVICE=
 EMAIL_USER=
 GDRIVE_CLIENT_EMAIL=
-GDRIVE_CLIENT_ID=
+GDRIVE_PROJECT_ID=
```
Comment (ChinemeremChigbo) [inline .env.sample:9]: I don't think there should be 2 GDRIVE_PROJECT_ID s

File: .eslintrc.json
Comment (ChinemeremChigbo) [inline .eslintrc.json:1]: I don't think this file needs to be modified, please revert.

File: docker-compose.yml
Comment (ChinemeremChigbo) [inline docker-compose.yml:1]: I don't think this file needs to be modified

File: .env.sample (hunk 9 above)
Comment (ChinemeremChigbo) [reply .env.sample:9]: I think this used to be GDRIVE_CLIENT_ID but GDRIVE_CLIENT_ID is not used anywhere in the code. I would just remove this line.
```suggestion
```

File: .env.sample [hunk 10]
```diff
@@ -6,11 +6,12 @@ EMAIL_PASS=
...
 GDRIVE_PRIVATE_KEY=
 GDRIVE_PROJECT_ID=
 POSTGRES_DATABASE=
 POSTGRES_PASSWORD=
 POSTGRES_PORT=
 POSTGRES_URL_NON_POOLING=
-POSTGRES_USER=
\ No newline at end of file
+POSTGRES_USER=
+VERCEL=
```
Comment (ChinemeremChigbo) [inline .env.sample:17]: VERCEL secret is unused. Please remove it.
```suggestion
```

File: .env.sample
Comment (ChinemeremChigbo) [inline .env.sample:1]: I think the only line that should change from the original file is the removal of `GDRIVE_CLIENT_ID` the rest of the file should remain the same.

File: .eslintrc.json [hunk 11]
```diff
@@ -1,12 +1,6 @@
...
-  "extends": [
-    "next",
-    "next/core-web-vitals",
-    "plugin:prettier/recommended"
-  ],
-  "plugins": [
-    "prettier"
-  ],
+  "extends": ["next", "next/core-web-vitals", "plugin:prettier/recommended"],
+  "plugins": ["prettier"],
```
Comment (ChinemeremChigbo) [inline .eslintrc.json:3]:
```suggestion
  "extends": [
    "next",
    "next/core-web-vitals",
    "plugin:prettier/recommended"
  ],
  "plugins": [
    "prettier"
  ],
```

File: .eslintrc.json [hunk 12]
```diff
@@ -16,4 +10,4 @@
     ],
     "no-unused-vars": "error"
   }
-}
\ No newline at end of file
+}
```
Comment (ChinemeremChigbo) [inline .eslintrc.json:13]: Revert

**Review by ChinemeremChigbo — CHANGES_REQUESTED** [review]
Reviewed first few files. Will review the remainder later.

File: app/api/deleteFile/route.ts [hunk 13]
```diff
@@ -0,0 +1,53 @@
...
+        { status: 500 }
+      );
+    }
+
+    const auth = new google.auth.GoogleAuth({
+      projectId: process.env.GDRIVE_PROJECT_ID,
+      scopes: 'https://www.googleapis.com/auth/drive',
+      credentials: {
+        type: 'service_account',
+        client_id: process.env.GDRIVE_PROJECT_ID,
```
Comment (ChinemeremChigbo) [inline app/api/deleteFile/route.ts:29]: Shouldn't this be the process.env.GDRIVE_CLIENT_ID? or is that not needed?

File: app/api/searchDrive/route.ts [hunk 14]
```diff
@@ -0,0 +1,54 @@
...
+      { status: 500 }
+    );
+  }
+
+  const auth = new google.auth.GoogleAuth({
+    projectId: process.env.GDRIVE_PROJECTID,
+    scopes: 'https://www.googleapis.com/auth/drive',
+    credentials: {
+      type: 'service_account',
+      client_id: process.env.GDRIVE_PROJECT_ID,
```
Comment (ChinemeremChigbo) [inline app/api/searchDrive/route.ts:25]: Should this be Client_ID?

File: app/api/uploadFile/route.ts [hunk 15]
```diff
@@ -0,0 +1,80 @@
...
+      { status: 500 }
+    );
+  }
+
+  const auth = new google.auth.GoogleAuth({
+    projectId: process.env.GDRIVE_PROJECTID,
+    scopes: 'https://www.googleapis.com/auth/drive',
+    credentials: {
+      type: 'service_account',
+      client_id: process.env.GDRIVE_PROJECT_ID,
```
Comment (ChinemeremChigbo) [inline app/api/uploadFile/route.ts:46]: Should this be client_id?

**Review by ChinemeremChigbo — APPROVED** [review]
Good work David! Please address my comments and then feel free to merge. Just a heads up that the actual upload UI in the sistema app won't need the listing functionality since the links to the files will be stored in the database, but thanks for the implementation regardless.

File: app/api/deleteFile/route.ts (hunk 13 above)
Comment (ludavidca) [reply app/api/deleteFile/route.ts:29]: Not needed, deleted and yeeted.

File: app/api/searchDrive/route.ts (hunk 14 above)
Comment (ludavidca) [reply app/api/searchDrive/route.ts:25]: Not needed, deleted and yeeted.

File: app/api/uploadFile/route.ts (hunk 15 above)
Comment (ludavidca) [reply app/api/uploadFile/route.ts:46]: Not needed, deleted and yeeted.

**Pull request merged**